Workflow failed - fault_injection_full

[Failed Run](https://github.com/microsoft/ebpf-for-windows/actions/runs/5204733810)
[Codebase](https://github.com/microsoft/ebpf-for-windows/tree/10735e577c676f6adebe04c35de6041bec0823d8)
Test name - `fault_injection_full`

```
Starting test *** jit_load_attach_detach_unload_random_v4_test ***
test threads per program    : 32
test duration (in minutes)  : 10
test verbose output         : false
test extension restart      : true
test extension restart delay: 250 ms
waiting on 96 test threads...
FATAL ERROR: OpenService failed. Service:netebpfext, Error: 1060
FATAL ERROR: Failed to restart extension: netebpfext
Error: Process completed with exit code 1.
```

--- a/libs/platform/CMakeLists.txt
+++ b/libs/platform/CMakeLists.txt
@@ -88,6 +88,7 @@ target_compile_definitions("platform_user" PRIVATE
 
 wdk_add_library("platform_kernel" STATIC WINVER "${EBPFFORWINDOWS_WDK_WINVER}"
   ${platform_sources}
+  kernel/ebpf_fault_injection_kernel.c
   kernel/ebpf_handle_kernel.c
   kernel/ebpf_platform_kernel.c
   kernel/ebpf_native_kernel.c

--- a/libs/platform/ebpf_object.c
+++ b/libs/platform/ebpf_object.c
@@ -228,8 +228,8 @@ void
 ebpf_object_tracking_terminate()
 {
     for (int index = 0; index < EBPF_COUNT_OF(_ebpf_id_table); index++) {
-        ebpf_assert(_ebpf_id_table[index].object == NULL || ebpf_fuzzing_enabled);
-        ebpf_assert(_ebpf_id_table[index].reference_count == 0 || ebpf_fuzzing_enabled);
+        ebpf_assert(_ebpf_id_table[index].object == NULL || usersim_fault_injection_is_enabled());
+        ebpf_assert(_ebpf_id_table[index].reference_count == 0 || usersim_fault_injection_is_enabled());
     }
 }
 

--- a/libs/platform/ebpf_platform.c
+++ b/libs/platform/ebpf_platform.c
@@ -4,8 +4,6 @@
 #include "ebpf_platform.h"
 #include "ebpf_tracelog.h"
 
-bool ebpf_fuzzing_enabled = false;
-
 static uint32_t _ebpf_platform_maximum_processor_count = 0;
 
 _Ret_range_(>, 0) uint32_t ebpf_get_cpu_count() { return _ebpf_platform_maximum_processor_count; }

--- a/libs/platform/ebpf_platform.h
+++ b/libs/platform/ebpf_platform.h
@@ -100,8 +100,6 @@ extern "C"
 
     typedef struct _ebpf_helper_function_addresses ebpf_helper_function_addresses_t;
 
-    extern bool ebpf_fuzzing_enabled;
-
     /**
      * @brief Initialize the eBPF platform abstraction layer.
      * @retval EBPF_SUCCESS The operation was successful.
@@ -1075,6 +1073,18 @@ extern "C"
      */
     ebpf_result_t
     ebpf_utf8_string_to_unicode(_In_ const ebpf_utf8_string_t* input, _Outptr_ wchar_t** output);
+
+    /**
+     * @brief Check if fault injection is enabled. (Fault injection
+     *        is not supported in kernel mode.)
+     *
+     * TODO(#2677): move this prototype to another project.
+     *
+     * @retval TRUE Fault injection is enabled. (User mode *ONLY*)
+     * @retval FALSE Fault injection is not enabled.
+     */
+    BOOLEAN
+    usersim_fault_injection_is_enabled();
 
 #ifdef __cplusplus
 }

--- a/libs/platform/kernel/ebpf_fault_injection_kernel.c
+++ b/libs/platform/kernel/ebpf_fault_injection_kernel.c
@@ -1,0 +1,13 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_platform.h"
+
+// TODO(#2677): Update this prototype.
+BOOLEAN
+usersim_fault_injection_is_enabled(void)
+{
+    // Kernel mode replacement (for call from usersim).
+    // As of now, fault injection is not supported in kernel mode.
+    return FALSE;
+}

--- a/libs/platform/kernel/platform_kernel.vcxproj
+++ b/libs/platform/kernel/platform_kernel.vcxproj
@@ -38,6 +38,7 @@
     <ClCompile Include="..\ebpf_serialize.c" />
     <ClCompile Include="..\ebpf_state.c" />
     <ClCompile Include="..\ebpf_trampoline.c" />
+    <ClCompile Include="ebpf_fault_injection_kernel.c" />
     <ClCompile Include="ebpf_handle_kernel.c" />
     <ClCompile Include="ebpf_native_kernel.c" />
     <ClCompile Include="ebpf_platform_kernel.c" />

--- a/libs/platform/kernel/platform_kernel.vcxproj.filters
+++ b/libs/platform/kernel/platform_kernel.vcxproj.filters
@@ -73,6 +73,9 @@
     <ClCompile Include="..\ebpf_platform.c">
       <Filter>Source Files</Filter>
     </ClCompile>
+    <ClCompile Include=".\ebpf_fault_injection_kernel.c">
+      <Filter>Source Files</Filter>
+    </ClCompile>
   </ItemGroup>
   <ItemGroup>
     <ClInclude Include="..\ebpf_epoch.h">

--- a/libs/platform/unit/platform_unit_test.cpp
+++ b/libs/platform/unit/platform_unit_test.cpp
@@ -96,9 +96,19 @@ class _test_helper
     bool state_initiated = false;
 };
 
+struct ebpf_hash_table_destroyer_t
+{
+    void
+    operator()(_In_opt_ _Post_invalid_ ebpf_hash_table_t* table)
+    {
+        ebpf_hash_table_destroy(table);
+    }
+};
+
+using ebpf_hash_table_ptr = std::unique_ptr<ebpf_hash_table_t, ebpf_hash_table_destroyer_t>;
+
 TEST_CASE("hash_table_test", "[platform]")
 {
-    ebpf_hash_table_t* table = nullptr;
     std::vector<uint8_t> key_1(13);
     std::vector<uint8_t> key_2(13);
     std::vector<uint8_t> key_3(13);
@@ -135,23 +145,30 @@ TEST_CASE("hash_table_test", "[platform]")
         .bucket_count = 1,
     };
 
-    REQUIRE(ebpf_hash_table_create(&table, &options) == EBPF_SUCCESS);
+    ebpf_hash_table_t* raw_ptr = nullptr;
+    REQUIRE(ebpf_hash_table_create(&raw_ptr, &options) == EBPF_SUCCESS);
+    ebpf_hash_table_ptr table(raw_ptr);
 
     // Insert first
     // Empty bucket case
     REQUIRE(
-        ebpf_hash_table_update(table, key_1.data(), data_1.data(), EBPF_HASH_TABLE_OPERATION_INSERT) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 1);
+        ebpf_hash_table_update(table.get(), key_1.data(), data_1.data(), EBPF_HASH_TABLE_OPERATION_INSERT) ==
+        EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 1);
 
     // Insert second
     // Existing bucket, no backup.
-    REQUIRE(ebpf_hash_table_update(table, key_2.data(), data_2.data(), EBPF_HASH_TABLE_OPERATION_ANY) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 2);
+    REQUIRE(
+        ebpf_hash_table_update(table.get(), key_2.data(), data_2.data(), EBPF_HASH_TABLE_OPERATION_ANY) ==
+        EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 2);
 
     // Insert third
     // Existing bucket, with backup.
-    REQUIRE(ebpf_hash_table_update(table, key_3.data(), data_3.data(), EBPF_HASH_TABLE_OPERATION_ANY) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 3);
+    REQUIRE(
+        ebpf_hash_table_update(table.get(), key_3.data(), data_3.data(), EBPF_HASH_TABLE_OPERATION_ANY) ==
+        EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 3);
 
     // Iterate through all keys.
     uint64_t cookie = 0;
@@ -163,13 +180,14 @@ TEST_CASE("hash_table_test", "[platform]")
     values.resize(count);
     // Bucket contains 3 keys, but we only have space for 2.
     // Should fail with insufficient buffer.
-    REQUIRE(ebpf_hash_table_iterate(table, &cookie, &count, keys.data(), values.data()) == EBPF_INSUFFICIENT_BUFFER);
+    REQUIRE(
+        ebpf_hash_table_iterate(table.get(), &cookie, &count, keys.data(), values.data()) == EBPF_INSUFFICIENT_BUFFER);
     REQUIRE(count == 3);
     keys.resize(count);
     values.resize(count);
     // Bucket contains 3 keys, and we have space for 3.
     // Should succeed.
-    REQUIRE(ebpf_hash_table_iterate(table, &cookie, &count, keys.data(), values.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_iterate(table.get(), &cookie, &count, keys.data(), values.data()) == EBPF_SUCCESS);
 
     // Verify that all keys are found.
     for (size_t index = 0; index < 3; index++) {
@@ -187,64 +205,63 @@ TEST_CASE("hash_table_test", "[platform]")
         }
     }
     // Verify that there are no more keys.
-    REQUIRE(ebpf_hash_table_iterate(table, &cookie, &count, keys.data(), values.data()) == EBPF_NO_MORE_KEYS);
+    REQUIRE(ebpf_hash_table_iterate(table.get(), &cookie, &count, keys.data(), values.data()) == EBPF_NO_MORE_KEYS);
     REQUIRE(keys_found == 0x7);
 
     // Find the first
-    REQUIRE(ebpf_hash_table_find(table, key_1.data(), &returned_value) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_find(table.get(), key_1.data(), &returned_value) == EBPF_SUCCESS);
     REQUIRE(memcmp(returned_value, data_1.data(), data_1.size()) == 0);
 
     // Find the second
-    REQUIRE(ebpf_hash_table_find(table, key_2.data(), &returned_value) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_find(table.get(), key_2.data(), &returned_value) == EBPF_SUCCESS);
     REQUIRE(memcmp(returned_value, data_2.data(), data_2.size()) == 0);
 
     // Find the third
-    REQUIRE(ebpf_hash_table_find(table, key_2.data(), &returned_value) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_find(table.get(), key_2.data(), &returned_value) == EBPF_SUCCESS);
     REQUIRE(memcmp(returned_value, data_2.data(), data_2.size()) == 0);
 
     // Replace the second
     memset(data_2.data(), '0x55', data_2.size());
     REQUIRE(
-        ebpf_hash_table_update(table, key_2.data(), data_2.data(), EBPF_HASH_TABLE_OPERATION_REPLACE) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 3);
+        ebpf_hash_table_update(table.get(), key_2.data(), data_2.data(), EBPF_HASH_TABLE_OPERATION_REPLACE) ==
+        EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 3);
 
     // Find the first
-    REQUIRE(ebpf_hash_table_find(table, key_1.data(), &returned_value) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_find(table.get(), key_1.data(), &returned_value) == EBPF_SUCCESS);
     REQUIRE(memcmp(returned_value, data_1.data(), data_1.size()) == 0);
 
     // Next key
-    REQUIRE(ebpf_hash_table_next_key(table, nullptr, returned_key.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_next_key(table.get(), nullptr, returned_key.data()) == EBPF_SUCCESS);
     REQUIRE(returned_key == key_1);
 
-    REQUIRE(ebpf_hash_table_next_key(table, returned_key.data(), returned_key.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_next_key(table.get(), returned_key.data(), returned_key.data()) == EBPF_SUCCESS);
     REQUIRE(returned_key == key_2);
 
-    REQUIRE(ebpf_hash_table_next_key(table, returned_key.data(), returned_key.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_next_key(table.get(), returned_key.data(), returned_key.data()) == EBPF_SUCCESS);
     REQUIRE(returned_key == key_3);
 
-    REQUIRE(ebpf_hash_table_next_key(table, returned_key.data(), returned_key.data()) == EBPF_NO_MORE_KEYS);
+    REQUIRE(ebpf_hash_table_next_key(table.get(), returned_key.data(), returned_key.data()) == EBPF_NO_MORE_KEYS);
     REQUIRE(returned_key == key_3);
 
     // Delete middle key
-    REQUIRE(ebpf_hash_table_delete(table, key_2.data()) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 2);
+    REQUIRE(ebpf_hash_table_delete(table.get(), key_2.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 2);
 
     // Delete not found
-    REQUIRE(ebpf_hash_table_delete(table, key_2.data()) == EBPF_KEY_NOT_FOUND);
-    REQUIRE(ebpf_hash_table_key_count(table) == 2);
+    REQUIRE(ebpf_hash_table_delete(table.get(), key_2.data()) == EBPF_KEY_NOT_FOUND);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 2);
 
     // Find not found
-    REQUIRE(ebpf_hash_table_find(table, key_2.data(), &returned_value) == EBPF_KEY_NOT_FOUND);
+    REQUIRE(ebpf_hash_table_find(table.get(), key_2.data(), &returned_value) == EBPF_KEY_NOT_FOUND);
 
     // Delete first key
-    REQUIRE(ebpf_hash_table_delete(table, key_1.data()) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 1);
+    REQUIRE(ebpf_hash_table_delete(table.get(), key_1.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 1);
 
     // Delete last key
-    REQUIRE(ebpf_hash_table_delete(table, key_3.data()) == EBPF_SUCCESS);
-    REQUIRE(ebpf_hash_table_key_count(table) == 0);
-
-    ebpf_hash_table_destroy(table);
+    REQUIRE(ebpf_hash_table_delete(table.get(), key_3.data()) == EBPF_SUCCESS);
+    REQUIRE(ebpf_hash_table_key_count(table.get()) == 0);
 }
 
 void

--- a/tests/end_to_end/netsh_test.cpp
+++ b/tests/end_to_end/netsh_test.cpp
@@ -10,7 +10,6 @@
 #include "netsh_test_helper.h"
 #include "platform.h"
 #include "test_helper.hpp"
-#include "usersim/../../src/fault_injection.h"
 
 #include <winsock2.h>
 #include <windows.h>

--- a/tests/end_to_end/test_helper.cpp
+++ b/tests/end_to_end/test_helper.cpp
@@ -13,7 +13,6 @@
 #include "helpers.h"
 #include "mock.h"
 #include "test_helper.hpp"
-#include "usersim/../../src/fault_injection.h"
 
 #include <chrono>
 #include <filesystem>
@@ -23,8 +22,6 @@
 #include <mutex>
 #include <sstream>
 using namespace std::chrono_literals;
-
-extern "C" bool ebpf_fuzzing_enabled;
 
 bool _ebpf_capture_corpus = false;
 
@@ -257,7 +254,7 @@ GlueCloseHandle(HANDLE object_handle)
     bool found = _duplicate_handles.dereference_if_found(handle);
     if (!found) {
         // No duplicates. Close the handle.
-        if (!(ebpf_api_close_handle(handle) == EBPF_SUCCESS || ebpf_fuzzing_enabled)) {
+        if (!(ebpf_api_close_handle(handle) == EBPF_SUCCESS || usersim_fault_injection_is_enabled())) {
             throw std::runtime_error("ebpf_api_close_handle failed");
         }
     }
@@ -417,7 +414,7 @@ _Requires_lock_not_held_(_service_path_to_context_mutex) static void _preprocess
             const ebpf_operation_load_native_module_request_t* request =
                 (ebpf_operation_load_native_module_request_t*)user_request;
             size_t service_name_length = ((uint8_t*)request) + request->header.length - (uint8_t*)request->data;
-            REQUIRE(((service_name_length % 2 == 0) || ebpf_fuzzing_enabled));
+            REQUIRE(((service_name_length % 2 == 0) || usersim_fault_injection_is_enabled()));
 
             std::wstring service_path;
             service_path.assign((wchar_t*)request->data, service_name_length / 2);

--- a/tests/netebpfext_unit/netebpfext_unit.cpp
+++ b/tests/netebpfext_unit/netebpfext_unit.cpp
@@ -6,7 +6,6 @@
 #include "catch_wrapper.hpp"
 #include "netebpf_ext_helper.h"
 #include "passed_test_log.h"
-#include "usersim/../../src/fault_injection.h"
 #include "watchdog.h"
 
 #include <map>

